### PR TITLE
Monitor metric APIs

### LINKED ISSUE #1445: Can Sermant support Prometheus monitoring?
### What would you like to be added?

It is hoped that Sermant will support Prometheus monitoring so that Sermant's core metrics as well as each plugin's custom metrics can be monitored through Grafana.

### Why is this needed?

1. Through integration with Grafana, we can use its rich data visualization capabilities to create intuitive dashboards that can help us quickly locate and solve problems.
2. Through monitoring metrics, monitoring alarms can be added to ensure the stability of sermant.

###Proposal proposal
Regarding the need for Sermant to support Prometheus monitoring, I am currently considering the following three options and hope to discuss them in the community in order to choose the most suitable one or combine it with other more suitable options.
#### Solution 1: Integrate Micrometer
Advantages: Micrometer provides a common interface integrated with a variety of monitoring systems (including Prometheus), has complete documentation, and is easy to get started.
Disadvantages: need to introduce additional dependencies

#### Option 2: Use Prometheus’ native API directly
Advantages: Directly use Prometheus native capabilities
Disadvantages: Completely bound to Prometheus

#### Option 3: Implement custom Metric interface
advantage:
- Provides a flexible and scalable way to customize metrics, easy to implement and maintain
- Compatible with a variety of surveillance systems

shortcoming:
- Additional code needs to be developed to handle indicator registration and exposure
- It may take time for the community to adapt to this new interface

### PR DESCRIPTION
**Pull request by zwmagic**
**What type of PR is this?**

Feature

**What this PR does / why we need it?**

It is hoped that Sermant will support Prometheus monitoring so that Sermant's core metrics as well as each plugin's custom metrics can be monitored through Grafana.

**Which issue(s) this PR fixes？**

Fixes #1445

**Does this PR introduce a user-facing change?**

No

Get Prometheus metric at http:// 127.0.0.1:47128/sermant/metrics

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/sermant-io/Sermant/issues) related with this PR before you start working on it.
- [x] Make sure you have squashed your change to one single commit.
- [ ] GitHub Actions works fine in this PR.

### CODE REVIEW

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Counter.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+/*
+ * Copyright (C) 2024-2024 Huawei Technologies Co., Ltd. All rights reserved.
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Counter.java:2]: modifying the copyright

Comment (chengyouling) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Counter.java]: License using 'Copyright (C) 2024-2024 Sermant Authors. All rights reserved'

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+
+    private double[] percentiles;
+
+    private Integer percentilePrecision;
+
+    private Double minimumExpectedValue;
+
+    private Double maximumExpectedValue;
+
+    private Duration expiry;
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: The name is inappropriate. It should be the validity period.

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+import java.time.Duration;
+
+/**
+ * DistributionStatisticConfig
+ *
+ * @author zwmagic
+ * @since 2024-08-17
+ */
+public class DistributionStatisticConfig {
+    private Boolean percentileHistogram;
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: Use prefixes such as is and should for Boolean variables to specify the logic to be determined.

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    public DistributionStatisticConfig publishPercentiles(double[] percentilesValue) {
+        this.percentiles = percentilesValue;
+        return this;
+    }
+
+    /**
+     * Adds histogram buckets used to generate aggregable percentile approximations in monitoring systems that have
+     * query facilities to do so (e.g. Prometheus' {@code histogram_quantile}, Atlas' {@code :percentiles}).
+     *
+     * @return This builder.
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: not builder

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * metric tags
+ *
+ * @author zwmagic
+ * @since 2024-08-16
+ */
+public final class Tags {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java]: MetricTag

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+            DistributionStatisticConfig distributionStatisticConfig) {
+        return metric.summary(metricName, tags, description, distributionStatisticConfig);
+    }
+
+    /**
+     * NoopMetric
+     *
+     * @since 2024-08-17
+     */
+    static class NoopMetric implements Metric {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java]: why called NoopMetric

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+import java.util.concurrent.TimeUnit;
+import java.util.function.ToDoubleFunction;
+
+/**
+ * metric
+ *
+ * @author zwmagic
+ * @since 2024-08-17
+ */
+public class Metrics {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java]: The name must be specific, for example, MetricManager.  In addition, what is the relationship between this class and metric service?

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java (hunk 3 above)
Comment (lilai23) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: @daizhenyu I think these are same field definitions as micrmeter which are wrappered to use in Sermant. Keep the same is ok

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java [hunk 8]
```diff
@@ -0,0 +1,206 @@
...
+     * approximations.
+     *
+     * @param percentilesValue Percentiles to compute and publish. The 95th percentile should be expressed as
+     * {@code 0.95}.
+     * @return This builder.
+     */
+    public DistributionStatisticConfig publishPercentiles(double[] percentilesValue) {
+        this.percentiles = percentilesValue;
+        return this;
+    }
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: Note that if you quote refer or copy codes from other open source project，you have to declar it in Copyright. It is very important

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java (hunk 2 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: these are same field definitions as micrmeter which are wrappered to use in Sermant

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java (hunk 5 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java]: Try to keep the same naming as micrometer tags

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java (hunk 6 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java]: Set the default implementation, the real implementation is in agentcore-implement

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java (hunk 7 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Metrics.java]: these are same field definitions as micrmeter which are wrappered to use in Sermant

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java (hunk 8 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/DistributionStatisticConfig.java]: ok

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * The MeterMetricService class manages Micrometer metrics and provides initialization and data retrieval
+ * functionalities.
+ *
+ * @author zwmagic
+ * @since 2024-08-19
+ */
+@SpiWeight(Integer.MIN_VALUE + 10)
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java]: Using the spi configuration file, classes are loaded in order. Do not use annotations

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java [hunk 10]
```diff
@@ -113,7 +118,7 @@ protected ServiceManager() {
     public static void initServices() {
         ServiceConfig serviceConfig = ConfigManager.getConfig(ServiceConfig.class);
         List<String> startServiceArray = new ArrayList<>();
-        for (final BaseService service : ServiceLoader.load(BaseService.class,
+        for (final BaseService service : SpiLoadUtils.loadAll(BaseService.class,
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java]: Adjust the sequence of loaded classes in the SPI configuration file.

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+    /**
+     * Adds a key-value pair to the current Tags object.
+     *
+     * @param key The key.
+     * @param value The value.
+     * @return The current Tags object, for chaining calls.
+     */
+    public Tags add(String key, Object value) {
+        if (value == null) {
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java]: The key can also be null.

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    public Timer timer(String metricName, Tags tags, String description) {
+        io.micrometer.core.instrument.Timer timer = io.micrometer.core.instrument.Timer.builder(metricName)
+                .tags(getMeterTags(tags)).description(description).register(registry);
+        return new MeterTimer(registry, timer);
+    }
+
+    @Override
+    public Summary summary(String metricName, Tags tags, String description,
+            DistributionStatisticConfig distributionStatisticConfig) {
+        Builder builder = DistributionSummary.builder(metricName)
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java]: summaryBuilder

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+     */
+    private Iterable<Tag> getMeterTags(Tags tags) {
+        if (tags == null || tags.getSize() == 0) {
+            return EMPTY;
+        }
+
+        List<Tag> result = Lists.newArrayListWithCapacity(tags.getSize());
+        for (Map.Entry<String, String> entry : tags.getTags().entrySet()) {
+            String value = entry.getValue();
+            result.add(Tag.of(entry.getKey(), value == null ? "" : value));
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java]: if key is null ？

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+        io.micrometer.core.instrument.Counter counter = io.micrometer.core.instrument.Counter.builder(metricName)
+                .tags(getMeterTags(tags))
+                .description(description)
+                .register(registry);
+        return new MeterCounter(counter);
+    }
+
+    @Override
+    public Gauge gauge(String metricName, Tags tags, String description) {
+        return new MeterGauge(registry, metricName, getMeterTags(tags), description);
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java]: Different metrics are constructed in different ways. Some data such as the transferred tag and description of the metric is in the metermetric method, and some is in the metric construction method（Gauge）.

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+                        this.getClass().getClassLoader()).stream()
+                .collect(Collectors.toMap(MeterRegistryProvider::getType, provider -> provider));
+        MetricConfig metricConfig = ConfigManager.getConfig(MetricConfig.class);
+        String type = metricConfig.getType();
+        meterRegistryProvider = providerMap.get(type);
+        if (meterRegistryProvider == null) {
+            throw new IllegalArgumentException("can not find metricRegistry provider by " + type);
+        }
+        Metrics.addRegistry(meterRegistryProvider.getRegistry());
+        Metric metric = new MeterMetric(meterRegistryProvider, metricConfig);
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java]: private static Metric metric = new NoopMetric();  I feel that NoopMetric () is unnecessary.

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/prometheus/PrometheusMeterRegistryProvider.java [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+
+import org.kohsuke.MetaInfServices;
+
+/**
+ * Prometheus meter registry provider
+ *
+ * @author zwmagic
+ * @since 2024-08-19
+ */
+@MetaInfServices(MeterRegistryProvider.class)
```
Comment (daizhenyu) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/prometheus/PrometheusMeterRegistryProvider.java]: The spi configuration file is recommended.

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java (hunk 10 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java:121]: spi 配置是根据字符串排序的，无法保证加载类的顺序，

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java (hunk 15 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java:62]: 主要是避免因为metric未初始化就调用Metrics API 报错的场景：
1. 在MeterMetricService未初始化之前；
2. 在 agent.service.metric.enablem = false时

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/prometheus/PrometheusMeterRegistryProvider.java (hunk 16 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/prometheus/PrometheusMeterRegistryProvider.java]: fixed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java (hunk 9 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java]: fixed

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java (hunk 10 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/ServiceManager.java]: fixed

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java (hunk 11 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/Tags.java]: fixed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java (hunk 12 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java]: fixed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java (hunk 13 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetric.java]: fixed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java (hunk 15 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metrics/MeterMetricServiceImpl.java]: fixed

**Review by daizhenyu — DISMISSED** [review]

File: sermant-agentcore/sermant-agentcore-implement/pom.xml [hunk 17]
```diff
@@ -173,6 +174,11 @@
             <artifactId>grpc-protobuf</artifactId>
             <version>${grpc.version}</version>
         </dependency>
+        <dependency>
+            <groupId>io.micrometer</groupId>
+            <artifactId>micrometer-registry-prometheus</artifactId>
+            <version>${micrometer.version}</version>
+        </dependency>
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/pom.xml:181]: you should add licence of new dependency under https://github.com/sermant-io/Sermant/tree/develop/LICENSES/vendor and add declaration in https://github.com/sermant-io/Sermant/blob/develop/NOTICE

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metric/MeterMetric.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        if (distributionStatisticConfig.getMaximumExpectedValue() != null) {
+            summaryBuilder.maximumExpectedValue(distributionStatisticConfig.getMaximumExpectedValue());
+        }
+        if (distributionStatisticConfig.getExpiry() != null) {
+            summaryBuilder.distributionStatisticExpiry(distributionStatisticConfig.getExpiry());
+        }
+        if (distributionStatisticConfig.getBufferLength() != null) {
+            summaryBuilder.distributionStatisticBufferLength(distributionStatisticConfig.getBufferLength());
+        }
+        DistributionSummary summary = summaryBuilder.register(registry);
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metric/MeterMetric.java]: Optional.ofNullable(distributionStatisticConfig.getPercentileHistogram()).ifPresent(summaryBuilder::publishPercentileHistogram);
 may be more clear

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/config/MetricConfig.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+
+    /**
+     * The metric type, currently supports prometheus.
+     */
+    private String type = "prometheus";
+
+    /**
+     * The maximum number of metrics.
+     */
+    private Integer maxTimeSeries = MAXIMUM_TIME_SERIES_VALUE;
```
Comment (lilai23) [inline sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/config/MetricConfig.java]: add this configs to config.yaml

File: sermant-agentcore/sermant-agentcore-implement/pom.xml (hunk 17 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/pom.xml:181]: fixed

File: sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metric/MeterMetric.java (hunk 18 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/metric/MeterMetric.java]: fixed

File: sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/config/MetricConfig.java (hunk 19 above)
Comment (zwmagic) [reply sermant-agentcore/sermant-agentcore-core/src/main/java/io/sermant/core/service/metric/config/MetricConfig.java]: fixed

**Review by lilai23 — APPROVED** [review]

**Review by daizhenyu — APPROVED** [review]

**Pull request merged**